pkg/storage/segments/store.go

```
@@ -46,16 +46,95 @@ type Store interface {
+	r              io.Reader
+	n              int // number of bytes read into thresholdBuf
+	thresholdBuf   []byte
+	totalReadBytes int
+}
+
+// NewPeekThresholdReader creates a new instance of PeekThresholdReader
+func NewPeekThresholdReader(r io.Reader) (pt *PeekThresholdReader) {
+	return &PeekThresholdReader{r: r, n: 0, thresholdBuf: nil, totalReadBytes: 0}
+}
+
+// Read initially reads bytes from the internal buffer, then continues
+// reading from the wrapped data reader. The number of bytes read `n`
+// is returned.
+func (pt *PeekThresholdReader) Read(p []byte) (n int, err error) {
+
+	// Case 1: if the total number of bytes read is greater than the number
+	// of bytes read into the thresholdBuf, then Read is called on the given
+	// byte slice.
+	if pt.totalReadBytes > pt.n {
+		return pt.r.Read(p)
+	}
+
+	thresholdBytesRemaining := pt.n - pt.totalReadBytes
+
+	// Case 2: if the length of the given byte slice p is less than or equal
+	// to the number of threshold bytes remaining to be read, then the slice
+	// of the threshold buffer from the end of totalReadBytes to the length
+	// of p is copied to p.
+	if len(p) <= thresholdBytesRemaining {
+		tmp := pt.thresholdBuf[pt.totalReadBytes : pt.totalReadBytes+len(p)]
+		copy(p, tmp)
+		pt.totalReadBytes += len(p)
+		return len(p), nil
+	}
+
+	// Case 3: The buffer tail slice is created then read.
+	// A slice of read bytes is copied to the given byte slice p.
+	tmp := pt.thresholdBuf[pt.totalReadBytes : pt.totalReadBytes+thresholdBytesRemaining]
+	bufTail := make([]byte, len(p)-thresholdBytesRemaining)
+	numTailBytes, err := pt.r.Read(bufTail)
+	if err != nil {
+		return 0, err
+	}
+	tmp = append(tmp, bufTail...)
+	copy(p, tmp)
+	n = thresholdBytesRemaining + numTailBytes
+	pt.totalReadBytes += n
+	return n, nil
+}
+
+// checkSize returns a bool to determine whether a reader's size
+// is inline-sized or not.
+func (pt *PeekThresholdReader) isInline(thresholdSize int) (inline bool, err error) {
```

This method should be public - in case we want to move the PeekThresholdReader to another package, or use it from another package.

Also, I would suggest to change the name to `IsLargerThan`. It would be semantically more correct to check if "the reader is larger than a certain threshold" than checking if "the reader is inline". It's the segment that can be inline, not the reader. The implementation should be adjusted accordingly.

---

Also we need to embed some sanity checks:
- IsLargeThan/IsInline can be called only once. Any subsequent call should return an error.
- IsLargeThan/IsInline can be called only before the first call to Read. If Read has already been called, this method should return an error.